src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,426 @@
+   C = companion_matrix(f)
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+
+   if n==1 return [diagonal_join([CP for i in 1:D])] end
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+   return h1,h2
+end
+
+# generators for centralizer of unipotent elements are described in:
+# Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
+
+# V = vector of integers of the dimensions of Jordan blocks
+# return the generators for the centralizers of the diagonal join of unipotent Jordan blocks of dimensions V
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::FinField, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive root in this case
+   idN = identity_matrix(F,n)
+
+   # L = multiset(V)
+   L=[[V[1],1]]
+   for i in 2:length(V)
+      if V[i]==L[length(L)][1]
+         L[length(L)][2] += 1
+      else
+         push!(L, [V[i],1])
+      end
+   end
+   listgens = MatElem[]
+
+   # generators for GL + internal diagonal blocks
+   pos=1
+   for l in L
+      v_g = isSL ? _gens_for_SL(l[2],F) : _gens_for_GL(l[2],F)
+      for x in v_g
+         z = block_matrix(l[2],l[2],[x[i,j]*identity_matrix(F,l[1]) for i in 1:l[2] for j in 1:l[2]])
+         z = insert_block(idN,z,pos,pos)
+         push!(listgens,z)
+      end
+      if l[1]>1
+         for i in 1:l[1]-1, j in 1:degree(F)
+            z = identity_matrix(F,l[1])
+            for k in 1:l[1]-i z[k,i+k]=_lambda^j end
+            z = insert_block(idN,z,pos,pos)
+            push!(listgens,z)
+         end
+      end
+      pos += l[1]*l[2]
+   end
+
+   # external diagonal blocks
+   pos=1
+   for i in 1:length(L)-1
+      pos += L[i][1]*L[i][2]
+      # block above diagonal
+      z = idN
+      for j in 1:L[i][1]
+         idx = pos-L[i][1]+j-1
+         z[idx,idx+L[i+1][1]] = 1
+      end
+      push!(listgens,z)
+      # block below diagonal
+      z = idN
+      for j in 1:L[i][1]
+         idx = pos+j-1
+         z[idx,idx-L[i][1]]=1
+      end
+      push!(listgens,z)
+   end
+
+   # cardinality
+   res = prod([_GL_order(l[2],F) for l in L])
+   exp = fmpz(0)
+   for i in 1:length(L)-1, j in i+1:length(L)
+      exp += L[i][1]*L[i][2]*L[j][2]
+   end
+   exp *= 2
+   exp += sum([(L[i][1]-1)*L[i][2]^2 for i in 1:length(L)])
+   res *= order(F)^exp
+
+   return listgens, res
+end
+
+
+# V = vector of integers of the dimensions of Jordan blocks
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+# does the same as above, but every entry is replaced by the corresponding block matrix in the field F/(f)
```

Comment:
Hrm, it's a bit frustrating to have all this code duplication. Surely there must be a way to avoid this... Perhaps not in this PR, but at some point...

Reply:
The fact is that the first function works with single entries, the second one works with matrix blocks. But I'll see if I can improve it.